frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx

```diff
@@ -538,6 +544,192 @@ describe("QueryBuilder", () => {
   });
 
   describe("unsaved changes warning", () => {
+    describe("editing models", () => {
+      describe("editing queries", () => {
+        it("shows custom warning modal when leaving edited query via SPA navigation", async () => {
+          const { history } = await setup({
+            card: TEST_MODEL_CARD,
+            initialRoute: "/home",
+          });
+
+          history.push(`/model/${TEST_MODEL_CARD.id}/query`);
+
+          await waitFor(() => {
+            expect(
+              screen.queryByTestId("loading-spinner"),
+            ).not.toBeInTheDocument();
+          });
+
+          const rowLimitInput = await within(
+            screen.getByTestId("step-limit-0-0"),
+          ).findByPlaceholderText("Enter a limit");
+
+          userEvent.click(rowLimitInput);
+          userEvent.type(rowLimitInput, "0");
+
+          await waitFor(() => {
+            expect(rowLimitInput).toHaveValue(10);
+          });
+
+          userEvent.tab();
+
+          history.goBack();
+
+          expect(
+            screen.getByText("Changes were not saved"),
+          ).toBeInTheDocument();
+          expect(
+            screen.getByText(
+              "Navigating away from here will cause you to lose any changes you have made.",
+            ),
+          ).toBeInTheDocument();
```

Comment:
I plan to introduce a utility function for this and the opposite assertion (in a separate PR) because it adds too much clutter and too many LOC.